src/censored.jl

```diff
@@ -0,0 +1,343 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The density function (or mass function for discrete `d`) of the censored distribution is
+```math
+f(x; d, l, u) = \\begin{cases}
+    P_d(x \\le l), & x \\le l \\\\
+    P_d(x),         & l < x < u \\\\
+    P_d(x \\ge u), & x \\ge u \\\\
+  \\end{cases},
+```
+where ``P_d(x)`` is the density (or mass) function of `d`, and ``P_d(x \\le b)`` is the
+cumulative distribution function of ``d`` evaluated at ``x = b``.
+If ``x`` is a random variable from ``d``, then `clamp(x, l, u)` is a random variable from
+its censored version. Note that this implies that even if ``d`` is continuous, its censored
+form has discrete mixture components at the bounds.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Union{Missing,Real}, u::Union{Missing,Real})
+    return Censored(d, l, u)
+end
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real, TL<:Union{T,Missing}, TU<:Union{T,Missing}} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::TL      # lower bound
+    upper::TU      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        _is_non_empty_interval(l, u) || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T, T, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::Missing, u::T) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, Missing, T}(d, l, u)
+    end
+    function Censored(d::UnivariateDistribution, l::T, u::Missing) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T, T, Missing}(d, l, u)
+    end
+end
+Censored(d::UnivariateDistribution, l::Real, u::Real) = Censored(d, Base.promote(l, u)...)
+
+function censored(d::Censored, l::Union{Missing,Real}, u::Union{Missing,Real})
+    lower = if l isa Missing
+        d.lower isa Missing ? missing : d.lower
+    else
+        d.lower isa Missing ? l : max(l, d.lower)
+    end
+    upper = if u isa Missing
+        d.upper isa Missing ? missing : d.upper
+    else
+        d.upper isa Missing ? u : min(u, d.upper)
+    end
+    return Censored(d.uncensored, lower, upper)
+end
```

Comment:
Would have to implement the following three cases:

```suggestion
function censored(d::Censored, l::T, u::T) where {T<:Real}
    return censored(d.uncensored, d.lower === missing ? l : max(l, d.lower), d.upper === missing ? u : min(u, d.upper))
end
function censored(d::Censored, ::Missing, u::Real)
    return censored(d.uncensored, d.lower, d.upper === missing ? u : min(u, d.upper))
end
function censored(d::Censored, l::Real, ::Missing)
    return censored(d.uncensored, d.lower === missing ? l : max(l, d.lower), d.upper)
end
```

In any case I think it makes sense to call `censored` in case there exist special implementations for some combinations (possibly not all).